File: phase-i-matlab/Aureas 1.6/fcc5.m
%%%% parte final de los MFCCs
%FCCs con escalamiento lineal

% f es un segmento (canto)  calculado con estos parámetros: f=spectrogram(signal(:,1),flattopwin(256),128,256,FS); %256,128,256

function [y]=fcc5(canto,nfiltros,nc,nframes)


    [a,b]=size(canto);

    div=nframes;

    w=floor(b/div); %5
    p=1;

    for k=1:w:w*div
        B(:,p)=sum(abs(canto(:,k:k+w-1)).^2,2); %energía para todos los bines
        p=p+1;
    end

    numfilt=nfiltros;
    
    
    if  a>=numfilt
        H=zeros(numfilt,a);
        wf=floor(a/numfilt);
        i=1;
        for k=1:wf:wf*numfilt
            H(i,:)= gaussmf(1:a,[wf/4 k+wf-1]);
            %H(i,k:k+wf-1)=1;
            i=i+1;
        end
    end

    FBE = H * B;
    
    N=nc;
    M=numfilt;
    
    dctm = @( N , M )( sqrt(2.0/M) * cos( repmat([0:N-1].',1,M) ...
                                           .* repmat(pi*([1:M]-0.5)/M,N,1) ) );
    
    
    DCT = dctm( N , M );
    CC =  DCT * log( FBE );  
    
    y = CC;
end

File: phase-ii-server-client/Aureas_App/fcc5.py
import numpy as np

def gaussfm(x, sig, c):
	return np.exp((-(x-c)**2)/(2*(sig**2)))

def fcc5(canto, nfiltros, nc, nframes):
	canto = np.array(canto)
	[a,b] = np.shape(canto)
	div = nframes
	w = np.int(np.floor(b/div))

	if (w != 0) & (a >= nfiltros):
		B = np.zeros((a,1))
		
		for k in np.arange(0,w*div,w):
			B = np.concatenate((B,np.matrix(np.sum(np.abs(canto[:,k:k+w])**2,1)).T), axis=1)

		B = B[:,1:]

		H = np.zeros((nfiltros,a))
		if a >= nfiltros:
			wf = np.int(np.floor(a/nfiltros))
			i = 0
			for k in np.arange(0,wf*nfiltros,wf):
				H[i,:] = gaussfm(np.arange(1,a+1), wf/4, k+wf)
				i+=1

		FBE = np.dot(H,B)
		N = nc
		M = nfiltros
		DCT = np.sqrt(2.0/M)*np.cos(np.array(np.tile(np.matrix(np.arange(0,N)).T,(1,M)))*np.array(np.tile(np.pi*(np.arange(1,M+1)-0.5)/M,(N,1))))
		CC = np.dot(DCT,np.log(FBE))
		y = CC
	else:
		y = []
	return y